x/evm/types/msg.go

```diff
@@ -336,41 +293,43 @@ func (msg *MsgEthereumTx) GetFrom() sdk.AccAddress {
 }
 
 // AsTransaction creates an Ethereum Transaction type from the msg fields
-func (msg MsgEthereumTx) AsTransaction() *ethtypes.Transaction {
+func (msg *MsgEthereumTx) AsTransaction() *ethtypes.Transaction {
+	tx := msg.Raw.Transaction
+	if tx != nil {
+		return tx
+	}
+
+	// fallback to legacy format
 	txData, err := UnpackTxData(msg.Data)
 	if err != nil {
 		return nil
```

Comment:
do we need to check nil when call AsTransaction

Reply:
For new transactions it won't return `nil`.
For legacy transactions, as long as it passes CheckTx it won't return `nil`, but CheckTx is not 100% safe, so it can return `nil` here in rare cases.
For rpc queries, as long as it don't crash the node, I think it's fine to panic in rare cases.